tools/webpack-core/src/plugins/CopyExternalsPlugin.ts

```diff
@@ -83,47 +91,32 @@ export class CopyExternalsPlugin {
     if (!packageJson.dependencies && !packageJson.optionalDependencies)
       return;
 
-    const dependencies = [...Object.keys(packageJson.dependencies || {}), ...Object.keys(packageJson.optionalDependencies || {})];
+    // Grab external package's dependencies and all of its dependencies and so on recursively
+    const depsFromRecursion = await resolveRecursively({
+      path: path.dirname(packageJsonPath),
+    }) as DependencyStruct;
+    await this.recurseDependencies(depsFromRecursion.dependencies, outputDir);
+  }
+  // TODO: Optimize recursion, too many awaits.
+  private async recurseDependencies(dependencies: DependencyStruct[], outputDir: string) {
+    if (dependencies.length === 0)
+      return;
     for (const dep of dependencies) {
-      if (!this._copiedPackages.has(dep)) {
-        await this.copyPackage(dep, outputDir, path.dirname(packageJsonPath));
-        this._copiedPackages.add(dep);
-      }
+      if (this._copiedPackages.has(dep.name))
+        continue;
+      await this.copyPackage(dep.name, outputDir, `${dep.path}/package.json`); // add package.json to end so that path.dirname in copyPackage gets proper directory instead of the dir above it.

```

Comment:
Why not just do `dirname` at line 84 instead of in `copyPackage()`?

Reply:
Changed